### ISSUE TITLE
Resource leak in SCQIndexing, TableDirectoryListing, others..?

### ISSUE THREAD

**Issue by epickrram** [OpenHFT/Chronicle-Queue#437]
Anything that uses a `LongValue` via `ValueOut.int64ForBinding` (or similar) will leak a reference to the underlying `BytesStore`. Any of the reference types such as `BinaryLongReference`, `BinaryLongArrayReference` should probably also be `ReferenceCounted` and cleaned up when they are no longer required.

**Comment by epickrram** [comment]
@JerryShea if you revert https://github.com/OpenHFT/Chronicle-Bytes/commit/4bb615a14a60b761aa2b3a83d12d129fe001235c and rebuild, then `SingleChronicleQueueTest.mappedSegmentsShouldBeUnmappedAsCycleRolls` will start failing due to the leaked `BinaryLong*References` in `SCQIndexing` and `TableDirectoryListing`. This is likely to be a problem anywhere that `Binary*Reference` classes are used.

**Comment by RobAustin** [comment]
@jerryShea thIs effects the stability of chronicle queue. So is relatively high priority.

**Comment by JerryShea** [comment]
@RobAustin yep - will get on to it Monday am

**Comment by RobAustin** [comment]
Thanks

**Comment by JerryShea** [comment]
I don't know if it does have a huge effect on stability, as each BytesStore is provided by the MappedFile, and I agree with Mark's reasoning that when the MappedFile is released, nothing else will be using the BytesStores, as the SingleChronicleQueueStore etc. will also have been released.

As all the XXXReference classes are owned by something else, I made them Closeable and so the owning object can close them _in most cases_. Sometimes the XXXReference is held in a thread local, and so in this case we have to fall back to Mark's mechanism of releasing the BytesStore in MappedFile.performRelease.
